### PR TITLE
New operators in RxSwift 2.0 in Playgrounds

### PR DESCRIPTION
**Pull request by carlosypunto**

### CODE REVIEW

**Comment by carlosypunto** [comment]
For now do not merge WIP.

TODO:
- using (can anybody help me?)
- timeout

**Comment by kzaher** [comment]
I think demonstrating `using` could be tricky, and `timeout` is still on it's way :)

I'll think about it.

**Comment by carlosypunto** [comment]
thanks Kruno

Really missing more operators (WIP):

using (I need help)
generate
toObservable
mapWithIndex
takeWhileWithIndex
sampleLatest
shareReplay
retryWhen
share
flatMapWithIndex
flatMapFirst
flatMapLatest
retryWhen
Timeout (If that's the case)

**Comment by kzaher** [comment]
Hi @carlosypunto ,

We've released 2.0 a while ago, the interface has stabilized and I think we now have more time to update playgrounds with new operators.

I think we would first need to update these examples to new 2.0 API.
I like examples like this
```
Observable.of("🔴","🔴","🔴","🔵","🔵","🔵")
```
It makes it easy to visually inspect the example. Are there other colors and shapes also? Maybe we could print real marble diagrams :)

I think we would also probably need to change some wording in introduction. I can make suggestions, but it would be best if native english speakers would proofread it.

File: Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift [hunk 1]
```diff
@@ -1,171 +1,114 @@
...
 
-## Observables
-The key to understanding RxSwift is in understanding the notion of Observables. Creating them, manipulating them, and subscribing to them in order to react to changes.
+### Basis
 
-## Creating and Subscribing to Observables
-The first step in understanding this library is in understanding how to create Observables. There are a number of functions available to make Observables.
-Creating an Observable is one thing, but if nothing subscribes to the observable, then nothing will come of it so both are explained simultaneously.
-*/
+The key to understanding RxSwift is in understanding the notion of Observables. Creating them, manipulating them, and subscribing to them in order to react to changes.
```
Comment (kzaher) [inline Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift]: `...the notion of Observables (observable sequences).`

 `... in order to start computation and receive sequence elements?`

I don't think changes are most important concept, I think it's maybe more natural to think in terms of producing sequence elements.

File: Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift [hunk 2]
```diff
@@ -1,171 +1,114 @@
...
+
+### Why use RxSwift?
 
 A vast majority of the code we write revolves around responding to external actions. When a user manipulates a control, we need to write an @IBAction to respond to that. We need to observe Notifications to detect when the keyboard changes position. We must provide blocks to execute when URL Sessions respond with data. And we use KVO to detect changes in variables.
+
 All of these various systems makes our code needlessly complex. Wouldn't it be better if there was one consistent system that handled all of our call/response code? Rx is such a system.
 
-## Observables
-The key to understanding RxSwift is in understanding the notion of Observables. Creating them, manipulating them, and subscribing to them in order to react to changes.
+### Basis
```
Comment (kzaher) [inline Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift]: I think this should be `Basics`

File: Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift [hunk 3]
```diff
@@ -1,171 +1,114 @@
...
-The first step in understanding this library is in understanding how to create Observables. There are a number of functions available to make Observables.
-Creating an Observable is one thing, but if nothing subscribes to the observable, then nothing will come of it so both are explained simultaneously.
-*/
+The key to understanding RxSwift is in understanding the notion of Observables. Creating them, manipulating them, and subscribing to them in order to react to changes.
 
-/*:
-### empty
-`empty` creates an empty sequence. The only message it sends is the `.Completed` message.
-*/
+The first step in understanding this library is in understanding how to create Observables. There are a number of [functions available to make Observables](Creating_Observables).
```
Comment (kzaher) [inline Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift]: Maybe also use ` ... create observable sequences.` It's really important that people think in terms of sequences.

I think this other part should probably be:
```
There is a number of [factory functions available](Creating_Observables).
```

File: Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift [hunk 4]
```diff
@@ -1,171 +1,114 @@
...
 
-example("empty") {
-    let emptySequence: Observable<Int> = empty()
+Creating an Observable is one thing, but if nothing subscribes to the observable, then nothing will come.
```
Comment (kzaher) [inline Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift]: Maybe:

```
Defining an observable sequence is first step, but element will be produced only after an observer subscribes to receive it's elements.
```

File: Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift [hunk 5]
```diff
@@ -1,171 +1,114 @@
...
 
-/*:
-### just
-`just` represents sequence that contains one element. It sends two messages to subscribers. The first message is the value of single element and the second message is `.Completed`.
-*/
+### Subscribing to Observables
 
-example("just") {
-    let singleElementSequence = just(32)
+There are several ways for subscribe to Observable.
```
Comment (kzaher) [inline Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift]:
```
... to observable sequence.
```

File: Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift [hunk 6]
```diff
@@ -1,171 +1,114 @@
...
+### subscribe
 */
 
-example("sequenceOf") {
-    let sequenceOfElements/* : Observable<Int> */ = sequenceOf(0, 1, 2, 3, 4, 5, 6, 7, 8, 9)
-    
-    let subscription = sequenceOfElements
-        .subscribe { event in
-            print(event)
+example("subscribe") {
```
Comment (kzaher) [inline Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift]:
```
example("subscribe to receive raw events") {
```

File: Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift [hunk 7]
```diff
@@ -1,171 +1,114 @@
...
+Creating an Observable is one thing, but if nothing subscribes to the observable, then nothing will come.
 
-    let subscription = emptySequence
-        .subscribe { event in
-            print(event)
-        }
-}
+### Subscription
 
+When subscribes to an Observable, it begin to emit a sequence of `emum`s of type `Event`. For this reason Observable are also known as Sequences.
```
Comment (kzaher) [inline Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift]: I think we should advise people to think in terms of sequence of elements, and not events. Terminal events are just here to do plumbing and not so important. Maybe something like

```
After subscription sequence will begin emitting `Event` enums.
```

File: Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift [hunk 8]
```diff
@@ -1,171 +1,114 @@
...
+enum Event<Element> {
+    case Next(Element)    // Next element is produced
+    case Error(ErrorType) // Sequence terminates with error
+    case Completed        // Sequence completes sucessfully
+}
+```
 
-example("never") {
-    let neverSequence: Observable<String> = never()
+If the Observable emit `Event.Next` can still send events, but if Observable emit `Event.Error` or `Event.Completed`, Observable's emission is interrupted forever.
```
Comment (kzaher) [inline Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift]: Maybe explain this as...
```
After a subscription is made, sequence can produce arbitrary number of `Next` events (sequence elements), but no other event will be received after `Error` or `Completed` are emitted.

Sequence grammar explains this more concisely.

`Next* (Error | Completed)?`
```

File: Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift [hunk 9]
```diff
@@ -1,171 +1,114 @@
...
 */
 
-example("toObservable") {
-    let sequenceFromArray = [1, 2, 3, 4, 5].toObservable()
-
-    let subscription = sequenceFromArray
-        .subscribe { event in
-            print(event)
-        }
+example("subscribe") {
```
Comment (kzaher) [inline Rx.playground/Pages/Introduction.xcplaygroundpage/Contents.swift]: example("subscribe with optional callbacks per event type")

**Comment by carlosypunto** [comment]
@kzaher I continue here https://github.com/carlosypunto/RxSwift/tree/NewOperatorsInPlaygrounds 
In next commit I've done proposals at you, please check: https://github.com/carlosypunto/RxSwift/commit/5791c9cf76cb1c51e92ea17e2ded52669a927c5d
I close this

**Pull request closed without merging**